lib/http2/connection.c

```diff
@@ -455,7 +482,7 @@ static int handle_incoming_request(h2o_http2_conn_t *conn, h2o_http2_stream_t *s
         return 0;
     }
 
-    if (stream->_req_body.body == NULL) {
+    if (stream->_req_body.body == NULL || (header_exists_map & H2O_HPACK_PARSE_HEADERS_PROTOCOL_EXISTS) != 0) {
```

Comment:
Added doc comment at https://github.com/h2o/h2o/pull/1768/commits/b393eccaf173109cceedefb986ea519ad1306640.

> as well as looking into how the states are used and if we have a chance in simplifying the design.
I think the current design is most simple, because we have to receive websocket connect request from client and reply 200 response like ordinal request-response flow. While exchanging websocket DATA frames, the stream state is `SEND_BODY`.